cEP-0020.md

```diff
@@ -0,0 +1,284 @@
+n = Order.objects.all().count()
+StatisticByDate.objects.record(
+    metric=Metric.objects.PULL_REQUEST_COUNT,
+    value=n,
+    Period=Period.LIFETIME)
+```
+
+### Gamification
+
+As we have already discuss the game elements involved in the gamification
+system, we will assign some points to each of the activities we will be
+tracking for a newcomer and then accordingly we will unlock the levels and
+provide them badges.
+
+#### Assign points to GitHub/GitLab activities
+
+For assigning points to GitHub/GitLab activities the main factor of decision
+will be the difficulty level of the issues.
+
+E.g.: If a newcomer has created a difficulty/newcomer issue we will give
+him `10` points, similarly we will increase `10` points according to the
+increase in difficulty levels.
+
+Similar concepts will be applied to the pull requests but we will also
+consider the number of Gitmate errors on the PR which will help us in finding
+that if the user has run coala locally before pushing his/her changes.
+
+For activities related to the comments, we will also check if the newcomer
+has made the review comment or the issue comment or its a simple comment on
+the PR.
+
+But this may lead to failure when the newcomers started doing comments just
+for the sake of earning points, to avoid this type of spam we will use some
+of the ideas from **Meta-review system**, i.e.: We will also check the reactions
+on the comments and if it is Thumbs Down we will not give any points to the
+newcomers.
+
+#### Assign points to Gitter activities
+
+It is obvious that tracking “No. of Gitter messages” and giving points for
+that will promote un-necessary spamming in the channel. So, to avoid that we
+can apply some extra rules on Gitter messages, we will divide our message model
+into two parts:
+
+* Good behavior message model
+* Bad behavior message model
+
+So, before importing Gitter messages to our models we can check in which
+category this message fall? We can do that by applying
+[basic sentiment analysis](http://fjavieralba.com/basic-sentiment-analysis-with-python.html)
+on the messages.
+
+Basically, we will define two dictionaries of good and bad messages.
+In the bad one, we will keep patterns like “please review [issue link]”,
+“updated [issue link]”, “have a look [issue link]”, “[mention] [issue link]”
+etc. So, that a newcomer doesn’t beg for review all the time.
+In the good one, we will have the pattern of questions so we will generally
+match “?” in the end. But what happens if a newcomer started asking irrelevant
+questions? So for that, we can also define a rule like a newcomer can only ask a
+certain number of questions in a day. Then we will track both the message model
+of the newcomer and assign some points to both separately and see which model
+metric has grown fast or slow while the whole newcomer process and not promote
+them until they have learned to learn things by themselves.
```

Comment:
This whole sentiment analysis approach needs to be revised IMO and should be clearly stated. Whatever you have mentioned isa little vague, no?
How about _Can you review this?_ ? Begging for a review + Includes a question mark.

Reply:
Yeah, that was just an idea from my side. Would be great if you could suggest me some patterns in both the category. :)

Reply:
will follow along during the implementation phase of this part. ;)